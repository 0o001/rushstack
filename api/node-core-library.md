[Home](./index) &gt; [@microsoft/node-core-library](./node-core-library.md)

## node-core-library package

Core libraries that every NodeJS toolchain project should use.

## Classes

|  <p>Class</p> | <p>Description</p> |
|  --- | --- |
|  <p>[Colors](./node-core-library.colors.md)</p> | <p><b><i>(BETA)</i></b> The static functions on this class are used to produce colored text for use with the node-core-library terminal.</p> |
|  <p>[ConsoleTerminalProvider](./node-core-library.consoleterminalprovider.md)</p> | <p><b><i>(BETA)</i></b></p> |
|  <p>[Executable](./node-core-library.executable.md)</p> | <p>The Executable class provides a safe, portable, recommended solution for tools that need to launch child processes.</p> |
|  <p>[FileDiffTest](./node-core-library.filedifftest.md)</p> | <p>Implements a unit testing strategy that generates output files, and then compares them against the expected input. If the files are different, then the test fails.</p> |
|  <p>[FileSystem](./node-core-library.filesystem.md)</p> | <p>The FileSystem API provides a complete set of recommended operations for interacting with the file system.</p> |
|  <p>[FileWriter](./node-core-library.filewriter.md)</p> | <p>API for interacting with file handles.</p> |
|  <p>[InternalError](./node-core-library.internalerror.md)</p> | <p>An `Error` subclass that should be thrown to report an unexpected state that may indicate a software defect. An application may handle this error by instructing the end user to report an issue to the application maintainers.</p> |
|  <p>[JsonFile](./node-core-library.jsonfile.md)</p> | <p>Utilities for reading/writing JSON files.</p> |
|  <p>[JsonSchema](./node-core-library.jsonschema.md)</p> | <p>Represents a JSON schema that can be used to validate JSON data files loaded by the JsonFile class.</p> |
|  <p>[LegacyAdapters](./node-core-library.legacyadapters.md)</p> | <p><b><i>(BETA)</i></b> Helper functions used when interacting with APIs that do not follow modern coding practices.</p> |
|  <p>[LockFile](./node-core-library.lockfile.md)</p> | <p>A helper utility for working with file-based locks. This class should only be used for locking resources across processes, but should not be used for attempting to lock a resource in the same process.</p> |
|  <p>[MapExtensions](./node-core-library.mapextensions.md)</p> | <p>Helper functions for working with the `Map<K, V>` data type.</p> |
|  <p>[PackageJsonLookup](./node-core-library.packagejsonlookup.md)</p> | <p>This class provides methods for finding the nearest "package.json" for a folder and retrieving the name of the package. The results are cached.</p> |
|  <p>[PackageName](./node-core-library.packagename.md)</p> | <p>Various functions for working with package names that may include scopes.</p> |
|  <p>[Path](./node-core-library.path.md)</p> | <p>Common operations for manipulating file and directory paths.</p> |
|  <p>[ProtectableMap](./node-core-library.protectablemap.md)</p> | <p>The ProtectableMap provides an easy way for an API to expose a `Map<K, V>` property while intercepting and validating any write operations that are performed by consumers of the API.</p> |
|  <p>[Sort](./node-core-library.sort.md)</p> | <p>Operations for sorting collections.</p> |
|  <p>[StringBuilder](./node-core-library.stringbuilder.md)</p> | <p>This class allows a large text string to be constructed incrementally by appending small chunks. The final string can be obtained by calling StringBuilder.toString().</p> |
|  <p>[Terminal](./node-core-library.terminal.md)</p> | <p><b><i>(BETA)</i></b> This class facilitates writing to a console.</p> |
|  <p>[Text](./node-core-library.text.md)</p> | <p>Operations for working with strings that contain text.</p> |

## Enumerations

|  <p>Enumeration</p> | <p>Description</p> |
|  --- | --- |
|  <p>[FileConstants](./node-core-library.fileconstants.md)</p> | <p>String constants for common filenames and parts of filenames.</p> |
|  <p>[FolderConstants](./node-core-library.folderconstants.md)</p> | <p>String constants for common folder names.</p> |
|  <p>[NewlineKind](./node-core-library.newlinekind.md)</p> | <p>Enumeration controlling conversion of newline characters.</p> |
|  <p>[PosixModeBits](./node-core-library.posixmodebits.md)</p> | <p>An integer value used to specify file permissions for POSIX-like operating systems.</p> |
|  <p>[TerminalProviderSeverity](./node-core-library.terminalproviderseverity.md)</p> | <p><b><i>(BETA)</i></b></p> |

## Interfaces

|  <p>Interface</p> | <p>Description</p> |
|  --- | --- |
|  <p>[IColorableSequence](./node-core-library.icolorablesequence.md)</p> | <p><b><i>(BETA)</i></b></p> |
|  <p>[IConsoleTerminalProviderOptions](./node-core-library.iconsoleterminalprovideroptions.md)</p> | <p><b><i>(BETA)</i></b> Options to be provided to a [ConsoleTerminalProvider](./node-core-library.consoleterminalprovider.md)</p> |
|  <p>[IExecutableResolveOptions](./node-core-library.iexecutableresolveoptions.md)</p> | <p><b><i>(BETA)</i></b> Options for Executable.tryResolve().</p> |
|  <p>[IExecutableSpawnSyncOptions](./node-core-library.iexecutablespawnsyncoptions.md)</p> | <p><b><i>(BETA)</i></b> Options for Executable.execute().</p> |
|  <p>[IFileSystemCopyFileOptions](./node-core-library.ifilesystemcopyfileoptions.md)</p> | <p>The options for FileSystem.copyFile()</p> |
|  <p>[IFileSystemCreateLinkOptions](./node-core-library.ifilesystemcreatelinkoptions.md)</p> | <p>The options for `FileSystem.createSymbolicLinkJunction()`<!-- -->, `createSymbolicLinkFile()`<!-- -->, `createSymbolicLinkFolder()`<!-- -->, and `createHardLink()`<!-- -->.</p> |
|  <p>[IFileSystemDeleteFileOptions](./node-core-library.ifilesystemdeletefileoptions.md)</p> | <p>The options for FileSystem.deleteFile()</p> |
|  <p>[IFileSystemMoveOptions](./node-core-library.ifilesystemmoveoptions.md)</p> | <p>The options for FileSystem.move()</p> |
|  <p>[IFileSystemReadFileOptions](./node-core-library.ifilesystemreadfileoptions.md)</p> | <p>The options for FileSystem.readFile()</p> |
|  <p>[IFileSystemReadFolderOptions](./node-core-library.ifilesystemreadfolderoptions.md)</p> | <p>The options for FileSystem.readFolder()</p> |
|  <p>[IFileSystemUpdateTimeParameters](./node-core-library.ifilesystemupdatetimeparameters.md)</p> | <p>The parameters for `updateTimes()`<!-- -->. Both times must be specified.</p> |
|  <p>[IFileSystemWriteFileOptions](./node-core-library.ifilesystemwritefileoptions.md)</p> | <p>The options for FileSystem.writeFile()</p> |
|  <p>[IFileWriterFlags](./node-core-library.ifilewriterflags.md)</p> | <p>Interface which represents the flags about which mode the file should be opened in.</p> |
|  <p>[IJsonFileSaveOptions](./node-core-library.ijsonfilesaveoptions.md)</p> | <p>Options for JsonFile.saveJsonFile()</p> |
|  <p>[IJsonFileStringifyOptions](./node-core-library.ijsonfilestringifyoptions.md)</p> | <p>Options for JsonFile.stringify()</p> |
|  <p>[IJsonSchemaErrorInfo](./node-core-library.ijsonschemaerrorinfo.md)</p> | <p>Callback function arguments for JsonSchema.validateObjectWithCallback();</p> |
|  <p>[IJsonSchemaFromFileOptions](./node-core-library.ijsonschemafromfileoptions.md)</p> | <p>Options for JsonSchema.fromFile()</p> |
|  <p>[IJsonSchemaValidateOptions](./node-core-library.ijsonschemavalidateoptions.md)</p> | <p>Options for JsonSchema.validateObject()</p> |
|  <p>[IPackageJson](./node-core-library.ipackagejson.md)</p> | <p>An interface for accessing common fields from a package.json file.</p> |
|  <p>[IPackageJsonDependencyTable](./node-core-library.ipackagejsondependencytable.md)</p> | <p>This interface is part of the IPackageJson file format. It is used for the "dependencies", "optionalDependencies", and "devDependencies" fields.</p> |
|  <p>[IPackageJsonLookupParameters](./node-core-library.ipackagejsonlookupparameters.md)</p> | <p>Constructor parameters for [PackageJsonLookup](./node-core-library.packagejsonlookup.md)</p> |
|  <p>[IPackageJsonScriptTable](./node-core-library.ipackagejsonscripttable.md)</p> | <p>This interface is part of the IPackageJson file format. It is used for the "scripts" field.</p> |
|  <p>[IPackageJsonTsdocConfiguration](./node-core-library.ipackagejsontsdocconfiguration.md)</p> | <p><b><i>(BETA)</i></b> This interface is part of the IPackageJson file format. It is used for the "tsdoc" field.</p> |
|  <p>[IParsedPackageName](./node-core-library.iparsedpackagename.md)</p> | <p>A package name that has been separated into its scope and unscoped name.</p> |
|  <p>[IParsedPackageNameOrError](./node-core-library.iparsedpackagenameorerror.md)</p> | <p>Result object returned by [PackageName.tryParse()](./node-core-library.packagename.tryparse.md)</p> |
|  <p>[IProtectableMapParameters](./node-core-library.iprotectablemapparameters.md)</p> | <p>Constructor parameters for [ProtectableMap](./node-core-library.protectablemap.md)</p> |
|  <p>[IStringBuilder](./node-core-library.istringbuilder.md)</p> | <p>An interface for a builder object that allows a large text string to be constructed incrementally by appending small chunks.</p> |
|  <p>[ITerminalProvider](./node-core-library.iterminalprovider.md)</p> | <p><b><i>(BETA)</i></b> Implement the interface to create a terminal provider. Terminal providers can be registered to a [Terminal](./node-core-library.terminal.md) instance to receive messages.</p> |

## Type Aliases

|  <p>Type Alias</p> | <p>Description</p> |
|  --- | --- |
|  <p>[callback](./node-core-library.callback.md)</p> |  |
|  <p>[ExecutableStdioMapping](./node-core-library.executablestdiomapping.md)</p> | <p><b><i>(BETA)</i></b> Typings for IExecutableSpawnSyncOptions.stdio.</p> |
|  <p>[ExecutableStdioStreamMapping](./node-core-library.executablestdiostreammapping.md)</p> | <p><b><i>(BETA)</i></b> Typings for one of the streams inside IExecutableSpawnSyncOptions.stdio.</p> |

