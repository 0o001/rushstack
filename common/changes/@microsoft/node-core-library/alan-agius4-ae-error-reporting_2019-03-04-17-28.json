{
  "changes": [
    {
      "packageName": "@microsoft/node-core-library",
      "comment": "Introduce an interface INodePackageJson for loading package.json files whose \"version\" field may be missing.",
      "type": "minor"
    }
  ],
  "packageName": "@microsoft/node-core-library",
  "email": "alan.agius4@users.noreply.github.com"
}