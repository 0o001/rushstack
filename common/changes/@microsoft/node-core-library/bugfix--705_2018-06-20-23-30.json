{
  "changes": [
    {
      "packageName": "@microsoft/node-core-library",
      "comment": "issue #705: fallback on linux to /proc/{n}/stat if 'ps -p 1 -o lstart' is not supported",
      "type": "patch"
    }
  ],
  "packageName": "@microsoft/node-core-library",
  "email": "jbrekelmans@users.noreply.github.com"
}