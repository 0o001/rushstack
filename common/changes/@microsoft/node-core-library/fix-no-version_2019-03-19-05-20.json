{
  "changes": [
    {
      "packageName": "@microsoft/node-core-library",
      "comment": "Add two new APIs PackageJsonLookup.loadNodePackageJson() and tryLoadNodePackageJsonFor() that return INodePackageJson",
      "type": "minor"
    }
  ],
  "packageName": "@microsoft/node-core-library",
  "email": "4673363+octogonz@users.noreply.github.com"
}