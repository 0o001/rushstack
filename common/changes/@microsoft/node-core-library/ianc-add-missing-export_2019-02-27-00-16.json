{
  "changes": [
    {
      "packageName": "@microsoft/node-core-library",
      "comment": "Include an enum that had been missing from the exports.",
      "type": "patch"
    }
  ],
  "packageName": "@microsoft/node-core-library",
  "email": "iclanton@users.noreply.github.com"
}