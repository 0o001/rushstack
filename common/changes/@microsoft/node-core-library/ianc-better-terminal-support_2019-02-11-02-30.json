{
  "changes": [
    {
      "packageName": "@microsoft/node-core-library",
      "comment": "Include support for text formatting in the Terminal API.",
      "type": "minor"
    }
  ],
  "packageName": "@microsoft/node-core-library",
  "email": "iclanton@users.noreply.github.com"
}