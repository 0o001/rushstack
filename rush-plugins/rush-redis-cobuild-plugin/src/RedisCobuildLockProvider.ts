// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { createClient } from '@redis/client';

import type { ICobuildLockProvider, ICobuildContext, ICobuildCompletedState } from '@rushstack/rush-sdk';
import type {
  RedisClientOptions,
  RedisClientType,
  RedisFunctions,
  RedisModules,
  RedisScripts
} from '@redis/client';

/**
 * The redis client options
 * @public
 */
export interface IRedisCobuildLockProviderOptions extends RedisClientOptions {}

const KEY_SEPARATOR: string = ':';
const COMPLETED_STATE_SEPARATOR: string = ';';

export class RedisCobuildLockProvider implements ICobuildLockProvider {
  private readonly _options: IRedisCobuildLockProviderOptions;

  private _redisClient: RedisClientType<RedisModules, RedisFunctions, RedisScripts>;
  private _lockKeyMap: WeakMap<ICobuildContext, string> = new WeakMap<ICobuildContext, string>();
  private _completedKeyMap: WeakMap<ICobuildContext, string> = new WeakMap<ICobuildContext, string>();

  public constructor(options: IRedisCobuildLockProviderOptions) {
    this._options = options;
    this._redisClient = createClient(this._options);
  }

  public async acquireLockAsync(context: ICobuildContext): Promise<boolean> {
    const { terminal } = context;
    const lockKey: string = this.getLockKey(context);
    const incrResult: number = await this._redisClient.incr(lockKey);
    const result: boolean = incrResult === 1;
    terminal.writeDebugLine(`Acquired lock for ${lockKey}: ${incrResult}, 1 is success`);
    if (result) {
      await this.renewLockAsync(context);
    }
    return result;
  }

  public async renewLockAsync(context: ICobuildContext): Promise<void> {
    const { terminal } = context;
    const lockKey: string = this.getLockKey(context);
    await this._redisClient.expire(lockKey, 30);
    terminal.writeDebugLine(`Renewed lock for ${lockKey}`);
  }

  public async releaseLockAsync(context: ICobuildContext): Promise<void> {
    const { terminal } = context;
    const lockKey: string = this.getLockKey(context);
    await this._redisClient.set(lockKey, 0);
    terminal.writeDebugLine(`Released lock for ${lockKey}`);
  }

  public async setCompletedStateAsync(
    context: ICobuildContext,
    state: ICobuildCompletedState
  ): Promise<void> {
    const { terminal } = context;
    const key: string = this.getCompletedStateKey(context);
    const value: string = this._serializeCompletedState(state);
    await this._redisClient.set(key, value);
    terminal.writeDebugLine(`Set completed state for ${key}: ${value}`);
  }

  public async getCompletedStateAsync(context: ICobuildContext): Promise<ICobuildCompletedState | undefined> {
    const key: string = this.getCompletedStateKey(context);
    let state: ICobuildCompletedState | undefined;
    const value: string | null = await this._redisClient.get(key);
    if (value) {
      state = this._deserializeCompletedState(value);
    }
    return state;
  }

  /**
   * Returns the lock key for the given context
   * Example: cobuild:v1:<contextId>:<cacheId>:lock
   */
  public getLockKey(context: ICobuildContext): string {
    const { version, contextId, cacheId } = context;
    let lockKey: string | undefined = this._lockKeyMap.get(context);
    if (!lockKey) {
      lockKey = ['cobuild', `v${version}`, contextId, cacheId, 'lock'].join(KEY_SEPARATOR);
      this._completedKeyMap.set(context, lockKey);
    }
    return lockKey;
  }

  /**
   * Returns the completed key for the given context
   * Example: cobuild:v1:<contextId>:<cacheId>:completed
   */
  public getCompletedStateKey(context: ICobuildContext): string {
    const { version, contextId, cacheId } = context;
    let completedKey: string | undefined = this._completedKeyMap.get(context);
    if (!completedKey) {
      completedKey = ['cobuild', `v${version}`, contextId, cacheId, 'completed'].join(KEY_SEPARATOR);
      this._completedKeyMap.set(context, completedKey);
    }
    return completedKey;
  }

  private _serializeCompletedState(state: ICobuildCompletedState): string {
    // Example: SUCCESS;1234567890
    // Example: FAILURE;1234567890
    return `${state.status}${COMPLETED_STATE_SEPARATOR}${state.cacheId}`;
  }

  private _deserializeCompletedState(state: string): ICobuildCompletedState | undefined {
    const [status, cacheId] = state.split(COMPLETED_STATE_SEPARATOR);
    return { status: status as ICobuildCompletedState['status'], cacheId };
  }
}
