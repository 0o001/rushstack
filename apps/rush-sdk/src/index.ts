// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import * as path from 'path';
import { Import, IPackageJson, PackageJsonLookup } from '@rushstack/node-core-library';

// eslint-disable-next-line
let rushLibModule: any = (global as any).___rush___rushLibModule;

if (rushLibModule === undefined) {
  const importingPath: string | undefined = module?.parent?.filename;
  if (importingPath !== undefined) {
    const callerPackageFolder: string | undefined =
      PackageJsonLookup.instance.tryGetPackageFolderFor(importingPath);

    if (callerPackageFolder !== undefined) {
      const callerPackageJson: IPackageJson = require(path.join(callerPackageFolder, 'package.json'));

      const RUSH_LIB_NAME: string = '@microsoft/rush-lib';

      // Does the caller properly declare a dependency on rush-lib?
      if (
        (callerPackageJson.dependencies && callerPackageJson.dependencies[RUSH_LIB_NAME] !== undefined) ||
        (callerPackageJson.devDependencies &&
          callerPackageJson.devDependencies[RUSH_LIB_NAME] !== undefined) ||
        (callerPackageJson.peerDependencies &&
          callerPackageJson.peerDependencies[RUSH_LIB_NAME] !== undefined)
      ) {
        // Try to resolve rush-lib from the caller's folder
        try {
          const rushLibModulePath: string = Import.resolveModule({
            modulePath: '@microsoft/rush-lib',
            baseFolderPath: callerPackageFolder
          });

          rushLibModule = require(rushLibModulePath);
        } catch (error) {
          // If we fail to resolve it, ignore the error
        }
      }
    }
  }
}

if (rushLibModule === undefined) {
  throw new Error('The "@rushstack/rush-sdk" package context has not been initialized.');
}

// Based on TypeScript's __exportStar()
for (const property in rushLibModule) {
  if (property !== 'default' && !exports.hasOwnProperty(property)) {
    // Based on TypeScript's __createBinding()
    Object.defineProperty(exports, property, {
      enumerable: true,
      get: function () {
        return rushLibModule[property];
      }
    });
  }
}
