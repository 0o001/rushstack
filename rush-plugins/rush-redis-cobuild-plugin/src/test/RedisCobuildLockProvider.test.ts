/* eslint-disable @typescript-eslint/no-floating-promises */
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { Terminal, ConsoleTerminalProvider } from '@rushstack/node-core-library';
import { ICobuildCompletedState, ICobuildContext, OperationStatus } from '@rushstack/rush-sdk';
import { IRedisCobuildLockProviderOptions, RedisCobuildLockProvider } from '../RedisCobuildLockProvider';

import * as redisAPI from '@redis/client';
import type { RedisClientType } from '@redis/client';

const terminal = new Terminal(new ConsoleTerminalProvider());

describe(RedisCobuildLockProvider.name, () => {
  let storage: Record<string, string | number> = {};
  beforeEach(() => {
    jest.spyOn(redisAPI, 'createClient').mockImplementation(() => {
      return {
        incr: jest.fn().mockImplementation((key: string) => {
          storage[key] = (Number(storage[key]) || 0) + 1;
          return storage[key];
        }),
        expire: jest.fn().mockResolvedValue(undefined),
        set: jest.fn().mockImplementation((key: string, value: string) => {
          storage[key] = value;
        }),
        get: jest.fn().mockImplementation((key: string) => {
          return storage[key];
        })
      } as unknown as RedisClientType;
    });
  });

  afterEach(() => {
    jest.resetAllMocks();
    storage = {};
  });

  function prepareSubject(): RedisCobuildLockProvider {
    return new RedisCobuildLockProvider({} as IRedisCobuildLockProviderOptions);
  }
  const context: ICobuildContext = {
    contextId: '123',
    cacheId: 'abc',
    version: 1,
    terminal
  };

  it('getLockKey works', () => {
    const subject: RedisCobuildLockProvider = prepareSubject();
    const lockKey: string = subject.getLockKey(context);
    expect(lockKey).toMatchSnapshot();
  });

  it('getCompletedStateKey works', () => {
    const subject: RedisCobuildLockProvider = prepareSubject();
    const completedStateKey: string = subject.getCompletedStateKey(context);
    expect(completedStateKey).toMatchSnapshot();
  });

  it('acquires lock success', async () => {
    const subject: RedisCobuildLockProvider = prepareSubject();
    const result: boolean = await subject.acquireLockAsync(context);
    expect(result).toBe(true);
  });

  it('acquires lock fails at the second time', async () => {
    const subject: RedisCobuildLockProvider = prepareSubject();
    const cobuildContext: ICobuildContext = {
      ...context,
      contextId: 'abc'
    };
    const result1: boolean = await subject.acquireLockAsync(cobuildContext);
    expect(result1).toBe(true);
    const result2: boolean = await subject.acquireLockAsync(cobuildContext);
    expect(result2).toBe(false);
  });

  it('releaseLockAsync works', async () => {
    const subject: RedisCobuildLockProvider = prepareSubject();
    expect(() => subject.releaseLockAsync(context)).not.toThrowError();
  });

  it('set and get completedState works', async () => {
    const subject: RedisCobuildLockProvider = prepareSubject();
    const cacheId: string = 'foo';
    const status: ICobuildCompletedState['status'] = OperationStatus.SuccessWithWarning;
    expect(() => subject.setCompletedStateAsync(context, { status, cacheId })).not.toThrowError();
    const actualState: ICobuildCompletedState | undefined = await subject.getCompletedStateAsync(context);
    expect(actualState?.cacheId).toBe(cacheId);
    expect(actualState?.status).toBe(status);
  });
});
