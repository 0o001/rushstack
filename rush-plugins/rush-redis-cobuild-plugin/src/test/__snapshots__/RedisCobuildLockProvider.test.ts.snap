// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RedisCobuildLockProvider getCompletedStateKey works 1`] = `"cobuild:v1:123:abc:completed"`;

exports[`RedisCobuildLockProvider getLockKey works 1`] = `"cobuild:v1:123:abc:lock"`;
